src/java.desktop/share/classes/sun/awt/geom/Curve.java

```diff
@@ -712,6 +712,72 @@ public static int rectCrossingsForCubic(int crossings,
         return crossings;
     }
 
+    /**
+     * Return the t values that correspond to possible extrema in a given cubic function.
+     * <p>
+     * If the coefficient of the t^3 is large then the polynomial is a cubic and up to
+     * two values may be returned. If that coefficient is zero then the polynomial
+     * is a quadratic and up to one value may be returned. But if that coefficient is
+     * small then this method considers the possibility it could be either, so in that
+     * scenario this method may return up to three values. For ex: if the leading
+     * coefficient is .000001 that might be because the polynomial really is a cubic, or
+     * it might be because of rounding error and the polynomial is basically a quadratic.
+     * </p>
+     *
+     * @param coefficients four coefficients for a cubic polynomial equation. The nth element in this array is
+     *                     the coefficient for (t^n).
+     * @param dest an array to store the t values in. This must be at least 3 elements.
+     * @return the number of t-values that were stored in dest. This will be between 0-3.
+     */
+    public static int findExtrema(double[] coefficients, double[] dest) {
+
+        int returnValue = 0;
+
+        if (coefficients[3] != 0.0) {
+            // evaluate this as a cubic, where:
+
+            // f(t) = c[3] * t^3 + c[2] * t^2 + c[1] * t + c[0]
+            // df/dt = 3 * c[3] * t^2 + 2 * c[2] * t + c[1]
+
+            // so we have a quadratic polynomial:
+            // df/dt = A * t^2 + B * t + C
+
+            // ... where:
+            // A = 3 * c[3]
+            // B = 2 * c[2]
+            // C = c[1]
+
+            double[] eqn = new double[]{ coefficients[1],
+                    2.0 * coefficients[2],
+                    3.0 * coefficients[3] };
+            returnValue = QuadCurve2D.solveQuadratic(eqn, dest);
+
+            if (returnValue < 0.0)
+                returnValue = 0;
+        }
+
+        if (coefficients[3] > -.01 && coefficients[3] < .01 && coefficients[2] != 0.0) {
```

Comment:
do not test coefficients[3] within 0.1 !
Always use QuadCurve2D.solveQuadratic() that handles the case coefficient(x^2) = 0.
Finally this method findExtrema() is only necessary if control points (x or y) are given to determine the dx , dy polynoms and return roots... I prefer moving this code directly in getBounds2D() to have a more efficient implementation (= 1 method with 1 loop) to avoid allocation of the array double[] eqn = new double[].